app/app.go

```diff
@@ -682,6 +682,11 @@ func NewWasmApp(
 	app.scopedIBCKeeper = scopedIBCKeeper
 	app.scopedTransferKeeper = scopedTransferKeeper
 	app.scopedWasmKeeper = scopedWasmKeeper
+
+	app.SnapshotManager().RegisterExtensions(
+		wasm.NewWasmSnapshotter(filepath.Join(wasmDir, "wasm", "state", "wasm")),
```

Comment:
There's probably a nicer way of doing this, but honestly this directory structure is buried so deep inside https://github.com/CosmWasm/cosmwasm/blob/be4d94a8d73ceab359226b8ea0a5b4be032a07e6/packages/vm/src/cache.rs#L105 so maybe it's better to leave it that way for now. Just a thought.

Reply:
Yeah, that top level wasm comes from wasmd from long ago.
I have thought of changing it, but it would be a very painful breaking change to all running nodes.

Reply:
Actually, notice `wasmDir := filepath.Join(homePath, "wasm")` on line 448 or so. That already contains the first wasm.

It should be `filepath.Join(wasmDir, "state", "wasm")`

Reply:
The file structure is private and does not provide any stability guarantees. There is not even a guarantee Wasm is stored as file on disk in the next version of cosmwasm-vm. The file system should not be accessed from wasmd. Instead there is `GetCode` and `Create` in wasmvm to load and store a Wasm blob.

Reply:
There's another "wasm" along the way
https://github.com/CosmWasm/wasmd/blob/3fb737147bff3358bd7f028c96cde8f6c191ace5/x/wasm/keeper/keeper.go#L99

Reply:
> The file system should not be accessed from wasmd. Instead there is GetCode and Create in wasmvm to load and store a Wasm blob.

This is a good point, we can iterate over the values in the iavl tree rather than filenames. That is cleaner. And it will also "force" the node restoring from snapshotting to do the precompiling of the wasm files upon syncing, not later, providing more stable execution times later.

Reply:
sounds legit!